overlord/configstate/configcore/homedirs_test.go

```diff
@@ -0,0 +1,218 @@
+		{"/home/missing", `path "/home/missing/" does not exist`},
+		{"/home/existingFile", `path "/home/existingFile/" is not a directory`},
+		// combine a valid path with an invalid one
+		{"/home/existingDir,/boot/invalid", `path "/boot/invalid/" uses reserved root directory "/boot/"`},
+	} {
+		err := configcore.Run(coreDev, &mockConf{
+			state: s.state,
+			conf: map[string]interface{}{
+				"homedirs": testData.homedirs,
+			},
+		})
+		c.Assert(err, ErrorMatches, testData.expectedError, Commentf("%v", testData.homedirs))
+	}
+}
+
+func (s *homedirsSuite) TestConfigureWriteFailure(c *C) {
+	restore := configcore.MockEnsureFileState(func(path string, state osutil.FileState) error {
+		return errors.New("some write error")
+	})
+	defer restore()
+	err := configcore.Run(coreDev, &mockConf{
+		state: s.state,
+		conf: map[string]interface{}{
+			"homedirs": "/home/existingDir",
+		},
+	})
+	c.Assert(err, ErrorMatches, "some write error")
+}
+
+func (s *homedirsSuite) TestConfigureUnchanged(c *C) {
+	// Prepare the "system-params" config file with the same contents we expect
+	// it to have after applying the configuration.
+	snapStateDir := dirs.SnapdStateDir(dirs.GlobalRootDir)
+	err := os.MkdirAll(snapStateDir, 0755)
+	c.Assert(err, IsNil)
+	configPath := filepath.Join(snapStateDir, "system-params")
+	err = ioutil.WriteFile(configPath, []byte("homedirs=/home/existingDir\n"), 0644)
+	c.Assert(err, IsNil)
+
+	// The AppArmor configuration must not be updated; install a mock handler
+	// to later verify that it was not called.
+	tunableUpdated := false
+	restore := configcore.MockApparmorUpdateHomedirsTunable(func(paths []string) error {
+		tunableUpdated = true
+		return nil
+	})
+	defer restore()
+
+	err = configcore.Run(coreDev, &mockConf{
+		state: s.state,
+		conf: map[string]interface{}{
+			"homedirs": "/home/existingDir",
+		},
+	})
+	c.Assert(err, IsNil)
+	c.Check(tunableUpdated, Equals, false)
+}
+
+func (s *homedirsSuite) TestConfigureApparmorTunableFailure(c *C) {
+	var homedirs []string
+	restore := configcore.MockApparmorUpdateHomedirsTunable(func(paths []string) error {
+		homedirs = paths
+		return errors.New("tunable error")
+	})
+	defer restore()
+	err := configcore.Run(coreDev, &mockConf{
+		state: s.state,
+		conf: map[string]interface{}{
+			"homedirs": "/home/existingDir/one,/home/existingDir/two",
+		},
+	})
+	c.Check(err, ErrorMatches, "tunable error")
+	c.Check(homedirs, DeepEquals, []string{"/home/existingDir/one", "/home/existingDir/two"})
+}
+
+func (s *homedirsSuite) TestConfigureApparmorReloadFailure(c *C) {
+	restore := configcore.MockApparmorReloadAllSnapProfiles(func() error {
+		return errors.New("reload error")
+	})
+	defer restore()
+	err := configcore.Run(coreDev, &mockConf{
+		state: s.state,
+		conf: map[string]interface{}{
+			"homedirs": "/home/existingDir",
+		},
+	})
+	c.Assert(err, ErrorMatches, "reload error")
+}
+
+func (s *homedirsSuite) TestConfigureHomedirsHappy(c *C) {
```

Comment:
are we checking ApparmorUpdateHomedirsTunable here? should we check the file that it writes?

Reply:
Good catch, indeed I wasn't! I've now added a check.